From: =?utf-8?b?T25kxZllaiBTdXLDvQ==?= <ondrej@debian.org>
Date: Mon, 24 Jan 2022 11:49:32 +0100
Subject: Pick UTC date for release_date variable (Closes: #1000893)

---
 configure.ac | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/configure.ac b/configure.ac
index 48415e5..334c54f 100644
--- a/configure.ac
+++ b/configure.ac
@@ -1253,7 +1253,7 @@ AM_CONDITIONAL([HAVE_XELATEX], [test "$XELATEX" != ":" && test "$LATEXMK" != ":"
 # Pull release date from CHANGES file last modification date
 # for reproducible builds
 #
-release_date=`date -r CHANGES +%Y-%m-%d`
+release_date=`date -u -r CHANGES +%Y-%m-%d`
 AC_SUBST([RELEASE_DATE], $release_date)
 
 #
