From: =?utf-8?b?T25kxZllaiBTdXLDvQ==?= <ondrej@sury.org>
Date: Tue, 21 May 2019 17:56:58 +0000
Subject: Exit the ./gen program on failed readdir() call

---
 lib/dns/gen-unix.h | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/lib/dns/gen-unix.h b/lib/dns/gen-unix.h
index 84c3e7f..16b7524 100644
--- a/lib/dns/gen-unix.h
+++ b/lib/dns/gen-unix.h
@@ -28,8 +28,10 @@
 
 #include <sys/types.h>          /* Required on some systems for dirent.h. */
 
+#include <errno.h>
 #include <dirent.h>
 #include <stdbool.h>
+#include <stdlib.h>
 #include <unistd.h>		/* XXXDCL Required for ?. */
 
 #include <isc/lang.h>
@@ -66,9 +68,15 @@ next_file(isc_dir_t *dir) {
 	dir->filename = NULL;
 
 	if (dir->handle != NULL) {
+		errno = 0;
 		dirent = readdir(dir->handle);
-		if (dirent != NULL)
+		if (dirent != NULL) {
 			dir->filename = dirent->d_name;
+		} else {
+			if (errno != 0) {
+				exit(1);
+			}
+		}
 	}
 
 	if (dir->filename != NULL)
