From: Debian DNS Packaging <pkg-dns-devel@lists.alioth.debian.org>
Date: Fri, 24 Nov 2017 16:26:53 +0000
Subject: _version

# Add extensions to version
---
 version | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/version b/version
index 64517d5..5375630 100644
--- a/version
+++ b/version
@@ -8,4 +8,4 @@ MINORVER=11
 PATCHVER=2
 RELEASETYPE=-P
 RELEASEVER=1
-EXTENSIONS=
+EXTENSIONS="$(dpkg-parsechangelog --file=../debian/changelog | sed -n '/^Version/s/[^-]*//p')-$(dpkg-vendor --query Vendor)"
